File: mobile/android/FUN/app/src/main/java/com/fun/app/ui/components/player/PlayerOverlay.kt
/**
 * FUN App - Player Overlay (Android)
 * Information and interaction buttons
 */

package com.fun.app.ui.components.player

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import com.fun.app.data.models.FeedEpisode
import com.fun.app.ui.theme.*

@Composable
fun PlayerOverlay(
    episode: FeedEpisode,
    playerManager: VideoPlayerManager,
    onLike: () -> Void,
    onComment: () -> Unit,
    onShare: () -> Unit
) {
    var isLiked by remember { mutableStateOf(false) }
    val progress by playerManager.progress
    val currentPosition by playerManager.currentPosition
    val duration by playerManager.duration
    
    Box(modifier = Modifier.fillMaxSize()) {
        // Bottom gradient for readability
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(300.dp)
                .align(Alignment.BottomCenter)
                .background(
                    Brush.verticalGradient(
                        colors = listOf(
                            Color.Transparent,
                            Color.Black.copy(alpha = 0.7f)
                        )
                    )
                )
        )
        
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(16.dp)
        ) {
            // Top info
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = episode.series.title,
                    style = MaterialTheme.typography.titleMedium,
                    color = Color.White
                )
                
                // Credits badge
                Surface(
                    shape = MaterialTheme.shapes.medium,
                    color = Surface.copy(alpha = 0.8f)
                ) {
                    Row(
                        modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp),
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Icon(
                            imageVector = Icons.Default.Star,
                            contentDescription = null,
                            tint = Color.Yellow,
                            modifier = Modifier.size(16.dp)
                        )
                        Text(
                            text = "150",
                            style = MaterialTheme.typography.labelMedium,
                            color = Color.White
                        )
                    }
                }
            }
            
            Spacer(modifier = Modifier.weight(1f))
            
            // Bottom content and actions
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.Bottom
            ) {
                // Left: Episode info
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                        Text(
                            text = "Episode ${episode.episode.episodeNum}",
                            style = MaterialTheme.typography.labelLarge,
                            color = Color.White
                        )
                        
                        Text(
                            text = episode.episode.title,
                            style = MaterialTheme.typography.titleLarge,
                            color = Color.White
                        )
                    }
                    
                    episode.episode.description?.let { desc ->
                        Text(
                            text = desc,
                            style = MaterialTheme.typography.bodyMedium,
                            color = Color.White.copy(alpha = 0.9f),
                            maxLines = 2
                        )
                    }
                    
                    // Progress bar
                    if (duration > 0) {
                        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                            LinearProgressIndicator(
                                progress = progress,
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .height(3.dp),
                                color = Color.White,
                                trackColor = Color.White.copy(alpha = 0.3f)
                            )
                            
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween
                            ) {
                                Text(
                                    text = formatTime(currentPosition),
                                    style = MaterialTheme.typography.labelSmall,
                                    color = Color.White.copy(alpha = 0.8f)
                                )
                                
                                Text(
                                    text = formatTime(duration),
                                    style = MaterialTheme.typography.labelSmall,
                                    color = Color.White.copy(alpha = 0.8f)
                                )
                            }
                        }
                    }
                }
                
                Spacer(modifier = Modifier.width(16.dp))
                
                // Right: Action buttons
                Column(
                    verticalArrangement = Arrangement.spacedBy(24.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    ActionButton(
                        icon = if (isLiked) Icons.Default.Favorite else Icons.Default.FavoriteBorder,
                        count = episode.series.likeCount ?: 0,
                        tint = if (isLiked) Color.Red else Color.White,
                        onClick = {
                            isLiked = !isLiked
                            onLike()
                        }
                    )
                    
                    ActionButton(
                        icon = Icons.Default.Face,
                        count = 0,
                        tint = Color.White,
                        onClick = onComment
                    )
                    
                    ActionButton(
                        icon = Icons.Default.Share,
                        count = null,
                        tint = Color.White,
                        onClick = onShare
                    )
                    
                    IconButton(
                        onClick = { /* More options */ },
                        modifier = Modifier
                            .size(44.dp)
                            .clip(CircleShape)
                            .background(Color.White.copy(alpha = 0.2f))
                    ) {
                        Icon(
                            imageVector = Icons.Default.MoreVert,
                            contentDescription = "More",
                            tint = Color.White
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun ActionButton(
    icon: ImageVector,
    count: Int?,
    tint: Color,
    onClick: () -> Unit
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        IconButton(
            onClick = onClick,
            modifier = Modifier
                .size(44.dp)
                .clip(CircleShape)
                .background(Color.White.copy(alpha = 0.2f))
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = tint,
                modifier = Modifier.size(24.dp)
            )
        }
        
        count?.let {
            Text(
                text = formatCount(it),
                style = MaterialTheme.typography.labelSmall,
                color = Color.White
            )
        }
    }
}

private fun formatTime(millis: Long): String {
    val seconds = (millis / 1000).toInt()
    val minutes = seconds / 60
    val remainingSeconds = seconds % 60
    return String.format("%d:%02d", minutes, remainingSeconds)
}

private fun formatCount(count: Int): String {
    return when {
        count >= 1_000_000 -> String.format("%.1fM", count / 1_000_000.0)
        count >= 1_000 -> String.format("%.1fK", count / 1_000.0)
        else -> count.toString()
    }
}

File: mobile/ios/FUN/FUN/Views/Components/VideoPlayer/PlayerOverlay.swift
/**
 * FUN App - Player Overlay
 * Info and interaction buttons overlaid on video
 */

import SwiftUI

struct PlayerOverlay: View {
    let episode: FeedEpisode
    @ObservedObject var playerManager: VideoPlayerManager
    let onLike: () -> Void
    let onComment: () -> Void
    let onShare: () -> Void
    
    @State private var isLiked = false
    @State private var likeCount: Int = 0
    @State private var viewerCount: Int = 0
    @State private var showLikeAnimation = false
    
    var body: some View {
        ZStack {
            // Gradient overlay for readability
            VStack {
                Spacer()
                LinearGradient(
                    colors: [.clear, .black.opacity(0.7)],
                    startPoint: .top,
                    endPoint: .bottom
                )
                .frame(height: 300)
            }
            .ignoresSafeArea()
            
            VStack {
                // Top info
                HStack {
                    Text(episode.series.title)
                        .font(.headline)
                        .foregroundColor(.white)
                    
                    Spacer()
                    
                    // Credits badge
                    HStack(spacing: 4) {
                        Image(systemName: "star.fill")
                            .font(.caption)
                            .foregroundColor(.yellow)
                        Text("150")
                            .font(.caption.weight(.semibold))
                            .foregroundColor(.white)
                    }
                    .padding(.horizontal, 10)
                    .padding(.vertical, 6)
                    .background(Colors.surface.opacity(0.8))
                    .cornerRadius(12)
                }
                .padding()
                
                Spacer()
                
                // Bottom content and actions
                HStack(alignment: .bottom) {
                    // Left: Episode info
                    VStack(alignment: .leading, spacing: 12) {
                        VStack(alignment: .leading, spacing: 4) {
                            Text("Episode \(episode.episode.episodeNum)")
                                .font(.subheadline.weight(.semibold))
                                .foregroundColor(.white)
                            
                            Text(episode.episode.title)
                                .font(.title3.weight(.bold))
                                .foregroundColor(.white)
                        }
                        
                        if let description = episode.episode.description {
                            Text(description)
                                .font(.subheadline)
                                .foregroundColor(.white.opacity(0.9))
                                .lineLimit(2)
                        }
                        
                        // Progress bar
                        if playerManager.duration > 0 {
                            VStack(spacing: 4) {
                                GeometryReader { geometry in
                                    ZStack(alignment: .leading) {
                                        Rectangle()
                                            .fill(.white.opacity(0.3))
                                            .frame(height: 3)
                                        
                                        Rectangle()
                                            .fill(.white)
                                            .frame(width: geometry.size.width * playerManager.progress, height: 3)
                                    }
                                    .cornerRadius(1.5)
                                }
                                .frame(height: 3)
                                
                                HStack {
                                    Text(formatTime(playerManager.currentTime))
                                        .font(.caption2)
                                        .foregroundColor(.white.opacity(0.8))
                                    
                                    Spacer()
                                    
                                    Text(formatTime(playerManager.duration))
                                        .font(.caption2)
                                        .foregroundColor(.white.opacity(0.8))
                                }
                            }
                        }
                    }
                    
                    Spacer()
                    
                    // Right: Action buttons
                    VStack(spacing: 24) {
                        // Like button
                        ActionButton(
                            icon: isLiked ? "heart.fill" : "heart",
                            count: episode.series.likeCount ?? 0,
                            color: isLiked ? .red : .white
                        ) {
                            isLiked.toggle()
                            onLike()
                        }
                        
                        // Comment button
                        ActionButton(
                            icon: "bubble.left",
                            count: 0,
                            color: .white
                        ) {
                            onComment()
                        }
                        
                        // Share button
                        ActionButton(
                            icon: "square.and.arrow.up",
                            count: nil,
                            color: .white
                        ) {
                            onShare()
                        }
                        
                        // More button
                        Button(action: {
                            // Show more options
                        }) {
                            Image(systemName: "ellipsis")
                                .font(.title2)
                                .foregroundColor(.white)
                                .frame(width: 44, height: 44)
                                .background(.ultraThinMaterial)
                                .clipShape(Circle())
                        }
                    }
                }
                .padding()
                .padding(.bottom, 20)
            }
        }
    }
    
    private func formatTime(_ seconds: Double) -> String {
        let minutes = Int(seconds) / 60
        let remainingSeconds = Int(seconds) % 60
        return String(format: "%d:%02d", minutes, remainingSeconds)
    }
}

struct ActionButton: View {
    let icon: String
    let count: Int?
    let color: Color
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            VStack(spacing: 6) {
                Image(systemName: icon)
                    .font(.title2)
                    .foregroundColor(color)
                    .frame(width: 44, height: 44)
                    .background(.ultraThinMaterial)
                    .clipShape(Circle())
                
                if let count = count {
                    Text(formatCount(count))
                        .font(.caption)
                        .fontWeight(.semibold)
                        .foregroundColor(.white)
                }
            }
        }
    }
    
    private func formatCount(_ count: Int) -> String {
        if count >= 1_000_000 {
            return String(format: "%.1fM", Double(count) / 1_000_000)
        } else if count >= 1_000 {
            return String(format: "%.1fK", Double(count) / 1_000)
        } else {
            return "\(count)"
        }
    }
}
